### PR TITLE
Improve integration with ZIO

### PR DESCRIPTION
**Pull request by jdegoes**

### CODE REVIEW

File: src/main/scala/zio/json/encoder.scala [hunk 1]
```diff
@@ -4,42 +4,75 @@ import scala.annotation._
...
 
-// convenient `.toJson` syntax
-object syntax {
-  implicit final class EncoderOps[A](private val a: A) extends AnyVal {
-    def toJson(implicit A: Encoder[A]): String = A.toJson(a, None)
+import zio.{ Chunk, Managed, Queue, UIO, ZIO, ZManaged }
+import zio.blocking._
+import zio.stream._
 
-    // Jon Pretty's better looking brother, but a bit slower
```
Comment (phderome) [inline src/main/scala/zio/json/encoder.scala:12]: !!

File: src/test/scala/zio/json/EncoderTest.scala [hunk 2]
```diff
@@ -1,62 +1,61 @@
...
+import zio.json._
 import TestUtils._
 import scalaprops._
 import Property.{ implies, prop, property }
 import scala.collection.mutable
 
 import utest._
-import zio.json.data.googlemaps._
-import zio.json.data.twitter._
+import testtestzio.json.data.googlemaps._
```
Comment (phderome) [inline src/test/scala/zio/json/EncoderTest.scala]: odd package name

File: src/test/scala/zio/json/data/GoogleMaps.scala [hunk 3]
```diff
@@ -1,20 +1,20 @@
-package zio.json.data.googlemaps
+package testtestzio.json.data.googlemaps
```
Comment (phderome) [inline src/test/scala/zio/json/data/GoogleMaps.scala]: odd package name. `testzio` no?

File: src/main/scala/zio/json/decoder.scala [hunk 4]
```diff
@@ -4,91 +4,183 @@ import scala.annotation._
...
+  /**
+   * A `JsonError` value describes the ways in which decoding could fail. This structure is used
+   * to facilitate human-readable error messages during decoding failures.
+   */
   sealed abstract class JsonError
   object JsonError {
-    def render(trace: List[JsonError]): String =
-      trace.reverse.map {
+    def render(trace: Chunk[JsonError]): String =
+      trace.map {
```
Comment (phderome) [inline src/main/scala/zio/json/decoder.scala:183]: good, you didn't forget to remove the reverse.

File: src/test/scala/zio/json/data/Twitter.scala [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package zio.json.data.twitter
+package testtestzio.json.data.twitter
```
Comment (phderome) [inline src/test/scala/zio/json/data/Twitter.scala]: similarly as for GoogleMaps.

**Review by phderome — COMMENTED** [review]
I'd suggest the combinator operators or any other significant user-visible change you introduced should be mentioned and highlighted to some extent in the `README.md` (and possibly the variance changes in `README.md` should be reverted unless the variance changes are coming back imminently).

Also in `README.md`, since you're making many docs changes now:
1) `s/wheres we really expect the raw number/whereas we really expect the raw number/` 
2) `s/usecase/use case/`
3) `s/mainsteam/mainstream/`

If Sam approves and you think benchmarks improved, he should rerun the benchmarks and update numbers in `README.md` pending any CI tool to make benchmark numbers homogenous among contributors.

File: README.md [hunk 6]
```diff
@@ -167,36 +167,36 @@ A popular alternative way to encode sealed traits:
...
 ## Manual Instances
 
-Sometimes it is easier to reuse an existing `json.Decoder` rather than generate a new one. This can be accomplished using convenience methods on the `json.Decoder` typeclass to *derive* new decoders:
+Sometimes it is easier to reuse an existing `JsonDecoder` rather than generate a new one. This can be accomplished using convenience methods on the `JsonDecoder` typeclass to *derive* new decoders:
 
 ```scala
-trait Decoder[A] {
-  def map[B](f: A => B): Decoder[B]
-  def emap[B](f: A => Either[String, B]): Decoder[B]
+trait JsonDecoder[+A] {
```
Comment (phderome) [inline README.md]: you reverted variance changes in code but not here.

File: src/main/scala/zio/json/decoder.scala [hunk 7]
```diff
@@ -4,91 +4,183 @@ import scala.annotation._
...
-  final def widen[B >: A]: Decoder[B]                 = self.asInstanceOf[Decoder[B]]
-  final def xmap[B](f: A => B, g: B => A): Decoder[B] = map(f)
-  final def map[B](f: A => B): Decoder[B] =
-    new Decoder[B] {
-      override def unsafeDecodeMissing(trace: List[JsonError]): B =
-        f(self.unsafeDecodeMissing(trace))
-      def unsafeDecode(trace: List[JsonError], in: RetractReader): B =
-        f(self.unsafeDecode(trace, in))
+  /**
+   * Attempts to decode a stream of characters into a single value of type `A`, but mayu fail with
```
Comment (phderome) [inline src/main/scala/zio/json/decoder.scala]: mayu -> may

File: src/main/scala/zio/json/encoder.scala [hunk 8]
```diff
@@ -4,42 +4,108 @@ import scala.annotation._
...
+        runtime <- ZIO.runtime[Any].toManaged_
+        queue   <- Queue.bounded[Chunk[Char]](1).toManaged_
+        writer <- ZManaged.fromAutoCloseable {
+                   ZIO.effectTotal {
+                     new java.io.BufferedWriter(new Writer {
+                       override def write(buffer: Array[Char], offset: Int, len: Int): Unit =
+                         runtime.unsafeRun(queue.offer(Chunk.fromArray(buffer).drop(offset).take(len)))
+
+                       override def close(): Unit =
+                         runtime.unsafeRun(queue.shutdown)
```
Comment (fsvehla) [inline src/main/scala/zio/json/encoder.scala]: Shutting down the queue will fail the Stream with an interrupt.

File: src/main/scala/zio/json/encoder.scala [hunk 9]
```diff
@@ -4,42 +4,108 @@ import scala.annotation._
...
+  final def encodeJsonStream(a: A, indent: Option[Int]): ZStream[Blocking, Nothing, Char] =
+    ZStream.unwrapManaged {
+      (for {
+        runtime <- ZIO.runtime[Any].toManaged_
+        queue   <- Queue.bounded[Chunk[Char]](1).toManaged_
+        writer <- ZManaged.fromAutoCloseable {
+                   ZIO.effectTotal {
+                     new java.io.BufferedWriter(new Writer {
+                       override def write(buffer: Array[Char], offset: Int, len: Int): Unit =
+                         runtime.unsafeRun(queue.offer(Chunk.fromArray(buffer).drop(offset).take(len)))
```
Comment (fsvehla) [inline src/main/scala/zio/json/encoder.scala]: The underlying array is modified after calling `Chunk.fromArray`, leading to data corruption

File: src/main/scala/zio/json/encoder.scala (hunk 8 above)
Comment (fsvehla) [reply src/main/scala/zio/json/encoder.scala]: I have a fix for this, pushing `Exit` + using `collectWhileSuccess`, shall I open it against this branch or are we getting it in and I’ll follow up?

Comment (jdegoes) [reply src/main/scala/zio/json/encoder.scala]: @fsvehla Please open against this branch, if that's ok! And thanks for spotting.

File: src/main/scala/zio/json/encoder.scala (hunk 9 above)
Comment (jdegoes) [reply src/main/scala/zio/json/encoder.scala]: Good catch!

File: src/main/scala/zio/json/encoder.scala (hunk 8 above)
Comment (fsvehla) [reply src/main/scala/zio/json/encoder.scala]: Done

File: src/main/scala/zio/json/encoder.scala (hunk 1 above)
Comment (jdegoes) [reply src/main/scala/zio/json/encoder.scala:12]: It still exists, it just got moved to the package object!

**Pull request merged**